Algorithms/Knapsack/BranchAndBoundNode.cs

```diff
@@ -0,0 +1,98 @@
+using Utilities.Exceptions;
+
+namespace Algorithms.Knapsack
+{
+    public class BranchAndBoundNode
+    {
+        // taken --> true = the item where index = level is taken, vice versa
+        private readonly bool taken;
+
+        // cumulativeWeight --> um of weight of item associated in each nodes starting from root to this node (only item that is taken)
+        private int cumulativeWeight;
+
+        // cumulativeValue --> sum of value of item associated in each nodes starting from root to this node (only item that is taken)
+        private double cumulativeValue;
+
+        // upperBound --> largest possible value after taking/not taking the item associated to this node (fractional)
+        private double upperBound;
+
+        // level --> level of the node in the tree structure
+        private int level;
+
+        // parent node
+        private BranchAndBoundNode? parent;
+
+        public BranchAndBoundNode()
+        {
+        }
+
+        public BranchAndBoundNode(int level, bool taken, BranchAndBoundNode parent)
+        {
+            this.level = level;
+            this.taken = taken;
+            this.parent = parent;
+        }
+
+        public int GetCumulativeWeight()
+        {
+            return cumulativeWeight;
+        }
+
+        public void SetCumulativeWeight(int cumulativeWeight)
+        {
+            this.cumulativeWeight = cumulativeWeight;
+        }
+
+        public double GetCumulativeValue()
+        {
+            return cumulativeValue;
+        }
+
+        public void SetCumulativeValue(double cumulativeValue)
+        {
+            this.cumulativeValue = cumulativeValue;
+        }
+
+        public double GetUpperBound()
+        {
+            return upperBound;
+        }
+
+        public void SetUpperBound(double upperBound)
+        {
+            this.upperBound = upperBound;
+        }
```

Comment:
Was this ported from Java? Please use auto properties https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/auto-implemented-properties

Reply:
This algorithm was implemented in Java in my previous project. Didn't know about auto-implemented properties. Will change it. Thanks.